File: solutions/JavaScript/count_complete_tree_nodes_level_calculation.js
/**
 * Definition for a binary tree node.
 */
class TreeNode {
    constructor(val = 0, left = null, right = null) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

/**
 * Count nodes in complete binary tree using level calculation.
 * For complete tree, if left height == right height, left is perfect.
 * Time: O(log² n), Space: O(log n) for recursion
 * @param {TreeNode} root
 * @return {number}
 */
function countNodes(root) {
    if (!root) return 0;

    // Calculate left and right heights
    let leftHeight = 0;
    let leftNode = root.left;
    while (leftNode) {
        leftHeight++;
        leftNode = leftNode.left;
    }

    let rightHeight = 0;
    let rightNode = root.right;
    while (rightNode) {
        rightHeight++;
        rightNode = rightNode.right;
    }

    if (leftHeight === rightHeight) {
        // Left subtree is perfect: 2^(h+1) - 1 nodes + root + recursively count right
        return (1 << (leftHeight + 1)) - 1 + countNodes(root.right);
    } else {
        // Right subtree is perfect: 2^h - 1 nodes + root + recursively count left
        return (1 << (rightHeight + 1)) - 1 + countNodes(root.left);
    }
}

// Test cases
const root1 = new TreeNode(1);
root1.left = new TreeNode(2);
root1.right = new TreeNode(3);
root1.left.left = new TreeNode(4);
root1.left.right = new TreeNode(5);

console.log("Example 1 node count:", countNodes(root1));  // 5

const single = new TreeNode(1);
console.log("Example 2 node count:", countNodes(single));  // 1

console.log("Example 3 node count:", countNodes(null));  // 0

File: solutions/Golang/count_complete_tree_nodes_level_calculation.go
package main

import "fmt"

/**
 * Definition for a binary tree node.
 */
type TreeNode struct {
    Val   int
    Left  *TreeNode
    Right *TreeNode
}

/**
 * Count nodes in complete binary tree using level calculation.
 * For complete tree, if left height == right height, left is perfect.
 * Time: O(log² n), Space: O(log n) for recursion
 */
func countNodes(root *TreeNode) int {
    if root == nil {
        return 0
    }

    // Calculate left and right heights
    leftHeight := 0
    leftNode := root.Left
    for leftNode != nil {
        leftHeight++
        leftNode = leftNode.Left
    }

    rightHeight := 0
    rightNode := root.Right
    for rightNode != nil {
        rightHeight++
        rightNode = rightNode.Right
    }

    if leftHeight == rightHeight {
        // Left subtree is perfect: 2^(h+1) - 1 nodes + root + recursively count right
        return (1 << (leftHeight + 1)) - 1 + countNodes(root.Right)
    } else {
        // Right subtree is perfect: 2^h - 1 nodes + root + recursively count left
        return (1 << (rightHeight + 1)) - 1 + countNodes(root.Left)
    }
}

func main() {
    fmt.Println("Example 1: Count complete tree nodes with level calculation")
}
